### DISCUSSION TITLE
Improve the error message when vcpkg database corrupted

### DISCUSSION THREAD

**Discussion by dannypike** [microsoft/vcpkg#20859] [Ideas]
While install the latest version of boost, I got the following error:

```
Installing package boost-gil[core]:x64-windows...
Elapsed time for package boost-gil:x64-windows: 9.3 s
Starting package 16/61: libffi:x64-windows
Building package libffi[core]:x64-windows...
open_for_write(D:\work\github\vcpkg\buildtrees\libffi\stdout-x64-windows.log): permission denied
```

I don't know why it got the permission denied error (I was afk at the time); perhaps the folder was locked by some other program, e.g. another build I was doing in another window. However, I cannot restart the build because I now get this error:

```
D:\work\github\vcpkg>vcpkg install boost
Computing installation plan...
Error: database corrupted. Package fontconfig:x64-windows is installed but dependency gettext:x64-windows is not.
Note: Updating vcpkg by rerunning bootstrap-vcpkg may resolve this failure.
```
rerunning bootstrap-vcpkg did't fix it, so what do I do now?

**Environment**
- OS: Windows 10
- Compiler: whatever comes with Visual Studio 16.11.4

**Expected behavior**
Please give some better instructions on how to recover from a corrupted database. Do I have to delete the whole of my VCPKG installation and start again from scratch? Or is there a better way that won't take all weekend? :)

**Comment by JonLiu1993** [comment]
@dannypike ,Could you please delete the whole of your vcpkg and reinstall?

**Comment by dannypike** [comment]
I didn't want to do that because it's already about 17 hours into a build! I spent the time deleting and reinstalling individual dependencies until I found the one that was corrupted.

However, this ticket is not really about "how do I recover?" but more about "please give me enough information so that I might be able to avoid deleting everything and starting again".

So, please could you extend the error reporting to tell me what dependency `vcpkg` was looking at when it encountered the corruption?

**Comment by dg0yt** [comment]
The binary artifacts are normally cached outside the vcpkg tree. So if no port and not tool changed, you can remove (or rename) `installed` and `packages`, and it should restore matching artifacts without reinstall - at least if none the ABI inputs changed.

**Comment by dannypike** [comment]
thanks for the suggestion but I didn't want to do that. I would be throwing away a lot of time. I did eventually trace it by manually working my way back up the build tree but it would be nice if `vcpkg` gave a little more information, rather than saying only "the database is corrupt".

It's similar to those horrible Windows file dialogs that tell you a file operation failed but won't tell you what file it is talking about! :)

**Comment by dg0yt** [comment]
> thanks for the suggestion but I didn't want to do that. I would be throwing away a lot of time.

The suggestion was made in order to avoid rebuilding, and renaming would have given you the chance to interrupt if it wouldn't work.

A possibly trigger for such problems is running out of disk space.

**Comment by dannypike** [comment]
OK, thank you. I'm still not sure how it would help in this particular case but I'll try it next time to see if it gives me any extra info. I still think that `vcpkg` should give a dependency tree or something rather than just a "I encountered a fatal problem somewhere in the 138 packages that you have just built, but I won't tell you where it was or what I was trying to do when it happened".

This is a developer tool, so there's no need to shield this user from the details. Details is what this user likes to have when an error happens. :)

**Comment by JonLiu1993** [comment]
@dannypike , Sorry for the inconvenience, you can take @dg0yt ’s suggestion and try。

**Comment by dannypike** [comment]
@JonLiu1993 Thank you for your reply, but you have misunderstood my point.

I would have liked `vcpkg` to help me trace the source of the problem. The error reporting for `vcpkg` is usually excellent, with back tracing etc to describe the context. In this case, it was not good. All that `vcpkg` told me is that there was a "problem".

Please look at this part of the code and, when it detects a "database corruption", output some context. I did manage to work my way backwards through about six projects, removing them and testing the build each time until I found the one that failed. This was time consuming and could be improved by you telling me where the corruption was detected, i.e. which modules / projects, perhaps a back trace or something.

I do not need a solution that says "delete everything and start again". I need some help to reduce the amount of work that doing so would cause me.

**Comment by JackBoosY** [comment]
Please do not use the same vcpkg copy to perform multiple installation processes.
Can you use powershell without administrator privileges to try to delete this file?
And please provide output information in your pwsh that includes additional installation options `--debug`.

**Comment by dannypike** [comment]
@JackBoosY Thank you for your advice to avoid doing multiple simultaneous builds within a single `vcpkg` installation. I don't usually do it, but this build was so big that I wanted to keep it running all of the time and starting multiple installs was a quick and easy way to try to do that. I won't do it again.

I don't know what file you're talking about, but I don't think it matters. I didn't need to delete anything to fix the build. I reversed the sequence of `vcpkg install` commands using `vcpkg remove` until it stopped complaining about a "corrupt database". I then restarted doing `vcpkg install`s again (about 10 of them). That took hours but it worked.

So please don't think that I am looking for more ways to analyse and fix the database corruption, because I am not. There isn't a problem any more and so fixing that is not what I am asking for, here. What I am asking for is ...

__Please improve the contextual information output when `vcpkg` detects what it thinks is a "corrupt database". If I had a list of packages or relationships that are considered to be "corrupt", I could have restricted the list of reinstalls that I performed. Doing that might have saved me a lot of time.__

**Comment by sonygod** [comment]
I got the same issue and fixed it by 

     vcpkg remove --outdated --recurse
